upf/upf.c

```diff
@@ -565,13 +575,242 @@ format_upf_encap_trace (u8 * s, va_list * args)
   return s;
 }
 
+void
+upf_fpath_stack_dpo (upf_forwarding_policy_t * p)
+{
+  fib_path_list_contribute_forwarding(p->fib_pl,
+                                      FIB_FORW_CHAIN_TYPE_UNICAST_IP4,
+                                      FIB_PATH_LIST_FWD_FLAG_COLLAPSE, &p->dpo);
+}
+
 /* *INDENT-OFF* */
 VLIB_PLUGIN_REGISTER () =
 {
   .version = VPP_BUILD_VER,
 };
 /* *INDENT-ON* */
 
+// ####################  dpo restacking vft ####################
+static void
+upf_policy_destroy (upf_forwarding_policy_t * fp_entry)
+{
+  upf_main_t *gtm = &upf_main;
+  /*
+   * this upf_fp should not be a sibling on the path list, since
+   * that was removed when the API config went
+   */
+  ASSERT (fp_entry->fib_sibling == ~0);
+  ASSERT (fp_entry->fib_pl == FIB_NODE_INDEX_INVALID);
+
+  hash_unset_mem (gtm->forwarding_policy_by_id, fp_entry->policy_id);
+  vec_free (fp_entry->policy_id);
+  vec_free (fp_entry->rpaths);
+  pool_put (gtm->upf_forwarding_policies, fp_entry);
+}
+
+static upf_forwarding_policy_t *
+upf_policy_get_from_fib_node (fib_node_t * node)
+{
+  return ((upf_forwarding_policy_t *) (((char *) node) -
+				       STRUCT_OFFSET_OF
+				       (upf_forwarding_policy_t, fib_node)));
+}
+
+/**
+ * Function definition to get a FIB node from its index
+ */
+static fib_node_t *
+upf_policy_fib_node_get (fib_node_index_t index)
+{
+  upf_main_t *gtm = &upf_main;
+  upf_forwarding_policy_t *p;
+
+  p = pool_elt_at_index (gtm->upf_forwarding_policies, index);
+  return (&p->fib_node);
+}
+
+/**
+ * Function definition to inform the FIB node that its last lock has gone.
+ */
+static void
+upf_policy_last_lock_gone (fib_node_t * node)
+{
+  upf_policy_destroy (upf_policy_get_from_fib_node (node));
+}
+
+/**
+ * Function definition to backwalk a FIB node -
+ * Here we will restack the new dpo to forward node.
+ */
+static fib_node_back_walk_rc_t
+upf_policy_back_walk (fib_node_t * node, fib_node_back_walk_ctx_t * ctx)
+{
+  upf_fpath_stack_dpo (upf_policy_get_from_fib_node (node));
+  return (FIB_NODE_BACK_WALK_CONTINUE);
+}
+
+const fib_node_vft_t upf_fp_vft = {
+  .fnv_get = upf_policy_fib_node_get,
+  .fnv_last_lock = upf_policy_last_lock_gone,
+  .fnv_back_walk = upf_policy_back_walk,
+};
+
+static clib_error_t *
+upf_policy_init (vlib_main_t * vm)
+{
+  upf_policy_fib_node_type = fib_node_register_new_type (&upf_fp_vft);
+  return (NULL);
+}
+
+VLIB_INIT_FUNCTION (upf_policy_init);
+
+u8 *
+format_upf_policy (u8 * s, va_list * args)
+{
+  upf_main_t *gtm = &upf_main;
+  upf_forwarding_policy_t *fp_entry =
+    va_arg (*args, upf_forwarding_policy_t *);
+
+  s = format (s, "upf:[%d]: policy:%v",
+	      fp_entry - gtm->upf_forwarding_policies, fp_entry->policy_id);
+  s = format (s, "\n ");
+  if (FIB_NODE_INDEX_INVALID == fp_entry->fib_pl)
+    {
+      s = format (s, "no forwarding");
+    }
+  else
+    {
+      s = fib_path_list_format (fp_entry->fib_pl, s);
+    }
+  return (s);
+}
+
+// Quick show method
```

Comment:
not sure this function (not a method) does show anything... misplaced comment? (besides it being C++ style)

Reply:
yes, it's outdated comment. Fixed